### ISSUE TITLE
[Bug] Memory leak in `Trainer`'s training loop on Linux systems

### ISSUE THREAD

**Issue by inafergra** [pasqal-io/perceptrain#22]
### Short description

When training a model using the `Trainer` class, the CPU RAM usage increases unbounded with each training epoch. This bug is specific to Linux systems (it has been reproduced in both local Linux machines and a SLURM managed Linux cluster).

![Image](https://github.com/user-attachments/assets/9ba1bef5-cfab-4c31-ac8b-30479a068b18)

### Steps/Code to reproduce
``` 
import psutil
import torch
from functools import reduce
from operator import add
import matplotlib.pyplot as plt

from qadence import QuantumCircuit, Z
from qadence import hamiltonian_factory, hea, feature_map
from qadence import QNN

from perceptrain import TrainConfig, Trainer, to_dataloader
from perceptrain.callbacks import Callback
from perceptrain.config import TrainConfig

Trainer.set_use_grad(True)

n_qubits = 8
fm = feature_map(n_qubits)
ansatz = hea(n_qubits=n_qubits, depth=n_qubits)
observable = hamiltonian_factory(n_qubits, detuning=Z)
circuit = QuantumCircuit(n_qubits, fm, ansatz)

model = QNN(circuit, observable, backend="pyqtorch", diff_mode="ad")
batch_size = 100
input_values = {"phi": torch.rand(batch_size, requires_grad=True)}
pred = model(input_values)

criterion = torch.nn.MSELoss()
optimizer = torch.optim.Adam(model.parameters(), lr=0.1)


# Data
def loss_fn(model: torch.nn.Module, data: torch.Tensor) -> tuple[torch.Tensor, dict]:
    x, y = data[0], data[1]
    out = model(x)
    loss = criterion(out, y)
    return loss, {}


fn = lambda x, degree: 0.05 * reduce(
    add, (torch.cos(i * x) + torch.sin(i * x) for i in range(degree)), 0.0
)
x = torch.linspace(0, 10, batch_size).reshape(-1, 1)
y = fn(x, 5)
train_dataloader = to_dataloader(x, y, batch_size=batch_size, infinite=True)


# Callback to measure CPU RAM usage
cpu_ram_list = []


class CalculateRAM(Callback):
    def run_callback(self, trainer, config, writer):
        cpu_ram_gb = psutil.virtual_memory()[3] / 1000000000
        print("RAM Used (GB):", cpu_ram_gb)
        cpu_ram_list.append(cpu_ram_gb)


ram_callback = CalculateRAM(on="train_epoch_start", called_every=1)
config = TrainConfig(
    max_iter=50,
    batch_size=batch_size,
    print_every=1,
    write_every=1,
    callbacks=[ram_callback],
)

# Train
trainer = Trainer(
    model,
    optimizer,
    config,
    loss_fn=loss_fn,
    train_dataloader=train_dataloader,
)
trainer.fit()

# Plot RAM
plt.clf()
plt.plot(cpu_ram_list, label="CPU RAM")
plt.xlabel("iterations")
plt.legend()
plt.show()
```

### Would you like to work on this issue?

None

**Comment by inafergra** [comment]
Hey! I found out a possible reason contributing to the leak. In this [line](https://github.com/pasqal-io/perceptrain/blob/ed813e0aede3067f10341c8a78495ea1135e042b/perceptrain/trainer.py#L441) the train losses + metrics are appended to a `train_losses` list with their corresponding gradients. I think there is no reason for the gradients to be tracked there, right? If I detach the tensors in the loss and metrics so that only the value of the tensor is appended and not the grad, the memory consumption is greatly reduced in my local Linux machine (although it still increases a bit, running it in Mac seems to still have a lower memory consumption):

![Image](https://github.com/user-attachments/assets/f86c4055-5456-46e0-b6e7-e6ec9d299b48)

However, I do not understand why would this cause an issue in Linux systems and not in Mac. There is probably more things going wrong under the hood, but not tracking grads there might help anyway

**Comment by mlahariya** [comment]
@inafergra - this is resolved, and we can close it, right?

**Comment by inafergra** [comment]
> [@inafergra](https://github.com/inafergra) - this is resolved, and we can close it, right?

Yes, it is fixed.
